src/component/Chart.jsx

```diff
@@ -0,0 +1,206 @@
+                    chartType: child.type.chartType
+                };
+            });
+
+        const childrenGroups = List(childrenProps)
+            .groupBy(ii => ii.chartType);
+
+        this.state = {
+            dimensions: props.dimensions
+                .map((dimensionName: string): Object => {
+                    const dimensionKey = `${dimensionName}Dimension`;
+                    const scaleKey = `${dimensionName}Scale`;
+                    const scaleTypeKey = `${dimensionName}ScaleType`;
+
+                    return {
+                        dimensionKey,
+                        dimensionName,
+                        scaleKey,
+                        scaleTypeKey,
+                        dimensions: childrenGroups
+                            .get('canvas')
+                            .map(ii => ii[dimensionKey] || props[dimensionKey])
+                            .toSet()
+                            .toArray()
+                    };
+                })
+        };
+
+        this.getChildProps = this.getChildProps.bind(this);
+        this.getCanvasSize = this.getCanvasSize.bind(this);
+        this.getAxisSize = this.getAxisSize.bind(this);
+    }
+    getCanvasSize(): Object {
+        const {width = 0, height = 0, padding = []} = this.props;
+        // $FlowBug: flow cant handle default assignment on array destructuring
+        const [top = 0, right = 0, bottom = 0, left = 0] = padding;
+
+        return {
+            width: Math.max(width - left - right, 0), // clamp negatives
+            height: Math.max(height - top - bottom, 0), // clamp negatives
+            outerWidth: width,
+            outerHeight: height,
+            top,
+            right,
+            bottom,
+            left
+        };
+    }
+    getAxisSize(axisType: string): Object {
+        const {top, right, bottom, left, width, height} = this.getCanvasSize();
+        switch(axisType) {
+            case 'top':
+                return {
+                    width,
+                    height: top
+                };
+
+            case 'right':
+                return {
+                    width: right,
+                    height
+                };
+
+            case 'bottom':
+                return {
+                    width,
+                    height: bottom
+                };
+
+            case 'left':
+                return {
+                    width: left,
+                    height
+                };
+
+            default:
+                return {};
+
+        }
+    }
+    getDefaultScale(dimension: Object): Function {
+        const {dimensionName, scaleTypeKey, dimensionKey, dimensions} = dimension;
+        switch(dimensionName) {
+            case 'x':
+                return (pp: Object): Function => {
+                    return d3Scale[pp[scaleTypeKey] || 'scaleLinear']()
+                        .domain(pp.data.rows.map(row => row.get(pp[dimensionKey])).toArray())
```

Comment:
Any reason that this passes in a whole array rather than going `data.min()` and / or `data.max()`? Those are calculated once on demand and memoized so should be way more performant, and aren't very expensive if all d3 needs are the lower and upper bounds. If this does need an array of lots of values then you're better off using `chartData.getColumnData(pp[dimensionKey]).toArray()` as it's also memoized.